torch/csrc/jit/codegen/cuda/runtime/helpers.cu

```diff
@@ -193,6 +209,20 @@ __device__ float threshold(float x, double t, double v) {
   return x <= t ? v : x;
 }
 
+__device__ std::complex<double> where(
+    bool c,
+    std::complex<double> a,
```

Comment:
dumb question, would we have mismatched complex<float> and complex<double> here? No idea if implicit type promotion works for complex.

Reply:
This is a good question.

On the one hand, the answer is: Yes, this type promotion works for `complex<double>` and `complex<float>`. See: https://github.com/llvm/llvm-project/blob/main/libcxx/include/complex#L85. Note that for a general type `T` and `X`, this conversion support is added in C++14, see https://github.com/llvm/llvm-project/blob/main/libcxx/include/complex#L27

On the other hand, PyTorch eager mode does not do type promotion on where, and some of its behavior is very weird and does not make any sense to me:
```python
import torch

ab = torch.zeros(5, dtype=torch.bool)
af = torch.zeros(5, dtype=torch.float)
ad = torch.zeros(5, dtype=torch.double)

# This fail as expected
torch.where(ab, af, ad)  # RuntimeError: expected scalar type float but found double

# This works as expected
torch.where(ab, ad, 3.0)

# WTF are you doing!
torch.where(ab, af, 3.0)  # RuntimeError: expected scalar type float but found double
```
Probably I should submit an issue to upstream discussion this problem.

Reply:
Thanks for filing the bug.

> torch.where(ab, af, 3.0)  # RuntimeError: expected scalar type float but found double

hmmmm. This is really surprising to me. since we did run into mismatched dtype for 2nd/3rd argument in opinfo test. Wondering how we ended up there..... I'm suspicious that complex<double> and complex<float> would also give us that. But let's pretend that it doesn't exist and we can run the tests after merging this one. :crossed_fingers: